flocker/control/test/test_protocol.py

```diff
@@ -693,25 +717,99 @@ def test_configuration_change(self):
         service.startService()
-        protocol = ControlAMP(Clock(), service)
-        protocol.makeConnection(StringTransport())
-        sent = []
-        self.patch_call_remote(sent, protocol=protocol)
+        server = LoopbackAMPClient(client.locator)
+        service.connected(server)
 
         service.configuration_service.save(TEST_DEPLOYMENT)
-        # Should only be one callRemote call.
-        (sent,) = sent
-        self.assertArgsEqual(
-            sent,
-            (
-                (ClusterStatusCommand,),
-                dict(
-                    configuration=TEST_DEPLOYMENT,
-                    state=DeploymentState(),
-                )
-            )
+
+        self.assertEqual(
+            dict(configuration=TEST_DEPLOYMENT, state=DeploymentState()),
+            dict(configuration=agent.desired, state=agent.actual),
+        )
+
+    def test_second_configuration_change_waits_for_first_acknowledgement(self):
+        """
+        A second configuration change is only transmitted after acknowledgement
+        of the first configuration change is received.
+        """
+        agent = FakeAgent()
+        client = AgentAMP(Clock(), agent)
+        service = build_control_amp_service(self)
+        service.startService()
+
+        configuration = service.configuration_service.get()
+        modified_configuration = arbitrary_transformation(configuration)
+
+        server = LoopbackAMPClient(client.locator)
+        delayed_server = DelayedAMPClient(server)
+        # Send first update
+        service.connected(delayed_server)
+        first_agent_desired = agent.desired
+
+        # Send second update
+        service.configuration_service.save(modified_configuration)
+        second_agent_desired = agent.desired
+
+        delayed_server.respond()
+        third_agent_desired = agent.desired
+
+        self.assertEqual(
+            dict(
+                first=configuration,
+                second=configuration,
+                third=modified_configuration,
+            ),
+            dict(
+                first=first_agent_desired,
+                second=second_agent_desired,
+                third=third_agent_desired,
+            ),
+        )
+
+    def test_third_configuration_change_supercedes_second(self):
+        """
+        A third configuration change completely replaces a second configuration
+        change if the first configuration change has not yet been acknowledged.
+        """
+        agent = FakeAgent()
+        client = AgentAMP(Clock(), agent)
+        service = build_control_amp_service(self)
+        service.startService()
+
+        configuration = service.configuration_service.get()
+        modified_configuration = arbitrary_transformation(configuration)
+        more_modified_configuration = arbitrary_transformation(modified_configuration)
+
+        server = LoopbackAMPClient(client.locator)
+        delayed_server = DelayedAMPClient(server)
```

Comment:
:question: It could just be me, but the variable names in these tests confused me for while....here are some of my difficulties:
 * `DelayedAMPClient` wraps `LoopbackAMPClient` but assigning to `server` makes it look like server is significant. Perhaps `DelayedAMPClient(wraps=LoopbackAMPClient(client.locator)` would be clearer.
 * `first_agent_desired`, `second_agent_desired` ... got me thinking that we were simulating multiple connected agents....instead of the state of a single agent at three points in time.....perhaps `agent_desired_1`, `agent_desired_2` etc might be clearer.
 * `service` could be `control_service` for the avoidance of doubt.
 * `delayed_server` could be `delayed_agent_endpoint` (or something)